cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogPipelineLoader.java

```diff
@@ -0,0 +1,234 @@
+        }
+      }
+    }), Predicates.notNull()));
+  }
+
+  /**
+   * Creates the {@link LogPipelineSpecification} representing the given log pipeline configuration URL.
+   */
+  private <T extends LoggerContext> LogPipelineSpecification<T> load(Provider<T> contextProvider,
+                                                                     URL configURL) throws JoranException {
+    // Create one AppenderContext per config file
+    T context = contextProvider.get();
+
+    CConfiguration pipelineCConf = CConfiguration.copy(cConf);
+    LogPipelineConfigurator configurator = new LogPipelineConfigurator(pipelineCConf);
+    configurator.setContext(context);
+
+    configurator.doConfigure(configURL);
+
+    // Check if the configuration has any error in it.
+    if (!new StatusChecker(context).isErrorFree(Status.ERROR)) {
+      List<Status> errors = new ArrayList<>();
+      for (Status status : context.getStatusManager().getCopyOfStatusList()) {
+        if (status.getEffectiveLevel() == Status.ERROR) {
+          errors.add(status);
+        }
+      }
+      throw new JoranException("Configuration failed " + errors);
+    }
+
+    // Default the pipeline name to the config file name if it is not set
+    if (context.getName() == null) {
+      String path = configURL.getPath();
+      int idx = path.lastIndexOf("/");
+      context.setName(idx < 0 ? path : path.substring(idx + 1));
+    }
+
+    byte[] checkpointPrefix = Bytes.toBytes(context.getName());
+    String prefixNum = configurator.getExecutionContext().getProperty(Constants.Logging.PIPELINE_CHECKPOINT_PREFIX_NUM);
+    if (prefixNum != null) {
+      try {
+        checkpointPrefix = Bytes.toBytes(Integer.parseInt(prefixNum));
+      } catch (NumberFormatException e) {
+        LOG.warn("Ignore invalid {} setting for pipeline in {}",
+                 Constants.Logging.PIPELINE_CHECKPOINT_PREFIX_NUM, configURL);
+      }
+    }
+
+    return new LogPipelineSpecification<>(configURL, context,
+                                          setupConfig(configurator, pipelineCConf), checkpointPrefix);
+  }
+
+  /**
+   * Copies overridable configurations from the pipeline configuration into the given {@link CConfiguration}.
+   */
+  private CConfiguration setupConfig(JoranConfigurator configurator, CConfiguration cConf) {
+    Context context = configurator.getContext();
+
+    // The list of properties that can be overridden per pipeline configuration
+    Set<String> keys = ImmutableSet.of(
+      Constants.Logging.PIPELINE_BUFFER_SIZE,
+      Constants.Logging.PIPELINE_EVENT_DELAY_MS,
+      Constants.Logging.PIPELINE_KAFKA_FETCH_SIZE,
+      Constants.Logging.PIPELINE_CHECKPOINT_INTERVAL_MS,
+      Constants.Logging.PIPELINE_LOGGER_CACHE_SIZE,
+      Constants.Logging.PIPELINE_LOGGER_CACHE_EXPIRATION_MS
+    );
+
+    for (String key : keys) {
+      context.putProperty(key, cConf.get(key));
+      cConf.set(key, configurator.getExecutionContext().subst("${" + key + "}"));
```

Comment:
The code in here is to copy all cconf with prefix `log.pipeline.<pipeline_name>` to the context and it has to be done before `doConfigure` is called so that these properties will be available to logback (hence can be used in pipeline logback xml).

After `doConfigurer`, inside the pipeline logback xml, some properties may be defined to override the default values of the one defined in the cConf, hence they got copied back to cConf.